subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/InstrumentArtifactTransform.java

```diff
@@ -0,0 +1,128 @@
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.internal.Pair;
+import org.gradle.internal.classpath.ClassData;
+import org.gradle.internal.classpath.ClasspathBuilder;
+import org.gradle.internal.classpath.ClasspathWalker;
+import org.gradle.internal.classpath.InstrumentingTransformer;
+import org.gradle.internal.classpath.TransformedClassPath;
+import org.gradle.internal.classpath.types.InstrumentingTypeRegistry;
+import org.gradle.internal.file.FileException;
+import org.gradle.internal.file.FileHierarchySet;
+import org.gradle.util.internal.GFileUtils;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+
+import static org.gradle.api.internal.initialization.transform.InstrumentArtifactTransform.InstrumentArtifactTransformParameters;
+
+@CacheableTransform
+public abstract class InstrumentArtifactTransform implements TransformAction<InstrumentArtifactTransformParameters> {
+
+    private static final Logger LOGGER = Logging.getLogger(InstrumentArtifactTransform.class);
+
+    private static final int BUFFER_SIZE = 8192;
+
+    public interface InstrumentArtifactTransformParameters extends TransformParameters {
+        @InputFiles
+        @PathSensitive(PathSensitivity.NAME_ONLY)
+        ConfigurableFileCollection getClassHierarchy();
+
+        @Internal
+        ConfigurableFileCollection getGlobalCacheLocations();
+    }
+
+    @InputArtifact
+    @PathSensitive(PathSensitivity.NAME_ONLY)
+    public abstract Provider<FileSystemLocation> getInput();
+
+    private File getInputAsFile() {
+        return getInput().get().getAsFile();
+    }
+
+    @Override
+    public void transform(TransformOutputs outputs) {
+        // TransformedClassPath.handleInstrumentingArtifactTransform depends on the order and this naming, we should make it more resilient in the future
+        String instrumentedJarName = getInput().get().getAsFile().getName().replaceFirst("\\.jar$", TransformedClassPath.INSTRUMENTED_JAR_EXTENSION);
+        File outputFile = outputs.file(instrumentedJarName);
+        FileHierarchySet globalCacheSet = FileHierarchySet.empty();
+        for (File file : getParameters().getGlobalCacheLocations()) {
+            globalCacheSet = globalCacheSet.plus(file);
+        }
+        if (globalCacheSet.contains(getInputAsFile().getAbsolutePath())) {
+            outputs.file(getInput());
+        } else {
+            File originalFile = outputs.file(getInputAsFile().getName());
+            GFileUtils.copyFile(getInputAsFile(), originalFile);
+        }
```

Comment:
🤔 We should not need to copy inputs to outputs. Transforms allow you to emit some of your inputs as outputs as is.